// @needle-tools/engine/lib/engine-components/CharacterController.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Quaternion, Ray, Vector3 } from "three";
import { Mathf } from "../engine/engine_math";
import { serializable } from "../engine/engine_serialization";
import { CapsuleCollider } from "./Collider";
import { Behaviour } from "./Component";
import { Rigidbody } from "./RigidBody";
import { Animator } from "./Animator";
import { RaycastOptions } from "../engine/engine_physics";
import { getWorldPosition } from "../engine/engine_three_utils";
export class CharacterController extends Behaviour {
    center = new Vector3(0, 0, 0);
    radius = .5;
    height = 2;
    _rigidbody = null;
    get rigidbody() {
        if (this._rigidbody)
            return this._rigidbody;
        this._rigidbody = this.gameObject.getComponent(Rigidbody);
        if (!this._rigidbody)
            this._rigidbody = this.gameObject.addNewComponent(Rigidbody);
        return this.rigidbody;
    }
    onEnable() {
        let rb = this.rigidbody;
        let collider = this.gameObject.getComponent(CapsuleCollider);
        if (!collider)
            collider = this.gameObject.addNewComponent(CapsuleCollider);
        // rb.isKinematic = true;
        collider.center.copy(this.center);
        collider.radius = this.radius;
        collider.height = this.height;
        this.gameObject.rotation.x = 0;
        this.gameObject.rotation.z = 0;
        rb.lockRotationX = true;
        rb.lockRotationY = true;
        rb.lockRotationZ = true;
        // TODO: this doesnt work yet
        // setInterval(()=>{
        //     this.rigidbody.isKinematic = !this.rigidbody.isKinematic;
        //     console.log(this.rigidbody.isKinematic);
        // }, 1000)
    }
    move(vec) {
        this.gameObject.position.add(vec);
    }
    _activeGroundCollisions = new Set();
    onCollisionEnter(col) {
        for (const contact of col.contacts) {
            // console.log(contact.normal);
            if (contact.normal.y > .1) {
                this._activeGroundCollisions.add(col);
                break;
            }
        }
    }
    onCollisionExit(col) {
        this._activeGroundCollisions.delete(col);
    }
    get isGrounded() { return this._activeGroundCollisions.size > 0; }
}
__decorate([
    serializable(Vector3)
], CharacterController.prototype, "center", void 0);
__decorate([
    serializable()
], CharacterController.prototype, "radius", void 0);
__decorate([
    serializable()
], CharacterController.prototype, "height", void 0);
export class CharacterControllerInput extends Behaviour {
    controller;
    movementSpeed = 2;
    rotationSpeed = 2;
    jumpForce = 1;
    animator;
    lookForward = true;
    _currentSpeed = new Vector3(0, 0, 0);
    _currentAngularSpeed = new Vector3(0, 0, 0);
    _temp = new Vector3(0, 0, 0);
    _jumpCount = 0;
    _currentRotation;
    awake() {
        this._currentRotation = new Quaternion();
    }
    update() {
        if (this.controller?.isGrounded) {
            this._jumpCount = 0;
            this.animator?.SetBool("doubleJump", false);
        }
        const forward = this.context.input.isKeyPressed("w");
        const backward = this.context.input.isKeyPressed("s");
        const rotateLeft = this.context.input.isKeyPressed("a");
        const rotateRight = this.context.input.isKeyPressed("d");
        const jump = this.context.input.isKeyDown(" ");
        // if (jumpDown) this._jumpDownTime = this.context.time.time;
        // const jumpUp = this.context.input.isKeyUp(" ");
        const step = forward ? 1 : 0 + backward ? -1 : 0;
        this._currentSpeed.z += step * this.movementSpeed * this.context.time.deltaTime;
        // if (!this.controller || this.controller.isGrounded)
        this.animator?.SetBool("running", step != 0);
        this.animator?.SetBool("jumping", this.controller?.isGrounded === true && jump);
        this._temp.copy(this._currentSpeed);
        this._temp.applyQuaternion(this.gameObject.quaternion);
        if (this.controller)
            this.controller.move(this._temp);
        else
            this.gameObject.position.add(this._temp);
        const rotation = rotateLeft ? 1 : 0 + rotateRight ? -1 : 0;
        this._currentAngularSpeed.y += Mathf.toRadians(rotation * this.rotationSpeed) * this.context.time.deltaTime;
        if (this.lookForward && Math.abs(this._currentAngularSpeed.y) < .01) {
            const forwardVector = this.context.mainCameraComponent.forward;
            forwardVector.y = 0;
            forwardVector.normalize();
            this._currentRotation.setFromUnitVectors(new Vector3(0, 0, 1), forwardVector);
            this.gameObject.quaternion.slerp(this._currentRotation, this.context.time.deltaTime * 10);
        }
        this.gameObject.rotateY(this._currentAngularSpeed.y);
        this._currentSpeed.multiplyScalar(1 - this.context.time.deltaTime * 10);
        this._currentAngularSpeed.y *= 1 - this.context.time.deltaTime * 10;
        if (this.controller && jump && this.jumpForce > 0) {
            let canJump = this.controller?.isGrounded;
            if (!this.controller?.isGrounded && this._jumpCount === 1) {
                canJump = true;
                this.animator?.SetBool("doubleJump", true);
            }
            if (canJump) {
                this._jumpCount += 1;
                // TODO: factor in mass
                const rb = this.controller.rigidbody;
                // const fullJumpHoldLength = .1;
                const factor = this._jumpCount === 2 ? 2 : 1; // Mathf.clamp((this.context.time.time - this._jumpDownTime), 0, fullJumpHoldLength) / fullJumpHoldLength;
                rb.applyImpulse(new Vector3(0, 1, 0).multiplyScalar(this.jumpForce * factor));
            }
        }
        if (this.controller) {
            // TODO: should probably raycast to the ground or check if we're still in the jump animation
            const verticalSpeed = this.controller?.rigidbody.getVelocity().y;
            if (verticalSpeed < -1) {
                if (!this._raycastOptions.ray)
                    this._raycastOptions.ray = new Ray();
                this._raycastOptions.ray.origin.copy(getWorldPosition(this.gameObject));
                this._raycastOptions.ray.direction.set(0, -1, 0);
                const currentLayer = this.layer;
                this.gameObject.layers.disableAll();
                this.gameObject.layers.set(2);
                const hits = this.context.physics.raycast(this._raycastOptions);
                this.gameObject.layers.set(currentLayer);
                if ((hits.length && hits[0].distance > 2 || verticalSpeed < -10)) {
                    this.animator?.SetBool("falling", true);
                }
            }
            else
                this.animator?.SetBool("falling", false);
        }
    }
    _raycastOptions = new RaycastOptions();
}
__decorate([
    serializable(CharacterController)
], CharacterControllerInput.prototype, "controller", void 0);
__decorate([
    serializable()
], CharacterControllerInput.prototype, "movementSpeed", void 0);
__decorate([
    serializable()
], CharacterControllerInput.prototype, "rotationSpeed", void 0);
__decorate([
    serializable()
], CharacterControllerInput.prototype, "jumpForce", void 0);
__decorate([
    serializable(Animator)
], CharacterControllerInput.prototype, "animator", void 0);
