// @needle-tools/engine/lib/engine-components/EventType.js
export var EventType;
(function (EventType) {
    /// <summary>
    /// Intercepts a IPointerEnterHandler.OnPointerEnter.
    /// </summary>
    EventType[EventType["PointerEnter"] = 0] = "PointerEnter";
    /// <summary>
    /// Intercepts a IPointerExitHandler.OnPointerExit.
    /// </summary>
    EventType[EventType["PointerExit"] = 1] = "PointerExit";
    /// <summary>
    /// Intercepts a IPointerDownHandler.OnPointerDown.
    /// </summary>
    EventType[EventType["PointerDown"] = 2] = "PointerDown";
    /// <summary>
    /// Intercepts a IPointerUpHandler.OnPointerUp.
    /// </summary>
    EventType[EventType["PointerUp"] = 3] = "PointerUp";
    /// <summary>
    /// Intercepts a IPointerClickHandler.OnPointerClick.
    /// </summary>
    EventType[EventType["PointerClick"] = 4] = "PointerClick";
    /// <summary>
    /// Intercepts a IDragHandler.OnDrag.
    /// </summary>
    EventType[EventType["Drag"] = 5] = "Drag";
    /// <summary>
    /// Intercepts a IDropHandler.OnDrop.
    /// </summary>
    EventType[EventType["Drop"] = 6] = "Drop";
    /// <summary>
    /// Intercepts a IScrollHandler.OnScroll.
    /// </summary>
    EventType[EventType["Scroll"] = 7] = "Scroll";
    /// <summary>
    /// Intercepts a IUpdateSelectedHandler.OnUpdateSelected.
    /// </summary>
    EventType[EventType["UpdateSelected"] = 8] = "UpdateSelected";
    /// <summary>
    /// Intercepts a ISelectHandler.OnSelect.
    /// </summary>
    EventType[EventType["Select"] = 9] = "Select";
    /// <summary>
    /// Intercepts a IDeselectHandler.OnDeselect.
    /// </summary>
    EventType[EventType["Deselect"] = 10] = "Deselect";
    /// <summary>
    /// Intercepts a IMoveHandler.OnMove.
    /// </summary>
    EventType[EventType["Move"] = 11] = "Move";
    /// <summary>
    /// Intercepts IInitializePotentialDrag.InitializePotentialDrag.
    /// </summary>
    EventType[EventType["InitializePotentialDrag"] = 12] = "InitializePotentialDrag";
    /// <summary>
    /// Intercepts IBeginDragHandler.OnBeginDrag.</
    /// </summary>
    EventType[EventType["BeginDrag"] = 13] = "BeginDrag";
    /// <summary>
    /// Intercepts IEndDragHandler.OnEndDrag.
    /// </summary>
    EventType[EventType["EndDrag"] = 14] = "EndDrag";
    /// <summary>
    /// Intercepts ISubmitHandler.Submit.
    /// </summary>
    EventType[EventType["Submit"] = 15] = "Submit";
    /// <summary>
    /// Intercepts ICancelHandler.OnCancel.
    /// </summary>
    EventType[EventType["Cancel"] = 16] = "Cancel";
})(EventType || (EventType = {}));
