// @needle-tools/engine/lib/engine-components/js-extensions/Object3D.js
import { applyPrototypeExtensions, registerPrototypeExtensions } from "./ExtensionUtils";
import { Object3D } from "three";
import { addNewComponent, getComponent, getComponentInChildren, getComponentInParent, getComponents, getComponentsInChildren, getComponentsInParent, getOrAddComponent, removeComponent } from "../../engine/engine_components";
import { isActiveSelf, setActive } from "../../engine/engine_gameobject";
// used to decorate cloned object3D objects with the same added components defined above
export function apply(object) {
    if (object && object.isObject3D === true) {
        applyPrototypeExtensions(object, Object3D);
    }
}
// do we still need this?
Object3D.prototype["SetActive"] = function (active) {
    this.visible = active;
};
Object3D.prototype["addNewComponent"] = function (type) {
    return addNewComponent(this, new type());
};
Object3D.prototype["removeComponent"] = function (inst) {
    return removeComponent(this, inst);
};
Object3D.prototype["getOrAddComponent"] = function (typeName) {
    return getOrAddComponent(this, typeName);
};
Object3D.prototype["getComponent"] = function (type) {
    return getComponent(this, type);
};
Object3D.prototype["getComponents"] = function (type, arr) {
    return getComponents(this, type, arr);
};
Object3D.prototype["getComponentInChildren"] = function (type) {
    return getComponentInChildren(this, type);
};
Object3D.prototype["getComponentsInChildren"] = function (type, arr) {
    return getComponentsInChildren(this, type, arr);
};
Object3D.prototype["getComponentInParent"] = function (type) {
    return getComponentInParent(this, type);
};
Object3D.prototype["getComponentsInParent"] = function (type, arr) {
    return getComponentsInParent(this, type, arr);
};
// this is a fix to allow gameObject active animation be applied to a three object
if (!Object.getOwnPropertyDescriptor(Object3D.prototype, "activeSelf")) {
    Object.defineProperty(Object3D.prototype, "activeSelf", {
        get: function () {
            return isActiveSelf(this);
        },
        set: function (val) {
            setActive(this, val, true);
        }
    });
}
// do this after adding the component extensions
registerPrototypeExtensions(Object3D);
