// @needle-tools/engine/src/engine-components/AvatarLoader.ts
import { GLTF, GLTFLoader } from "three/examples/jsm/loaders/GLTFLoader";
import * as THREE from "three";
import * as utils from "../engine/engine_utils"
// import * as object from "../engine/engine_gltf_builtin_components";
import * as loaders from "../engine/engine_loaders"
import { Context } from "../engine/engine_setup";
import { GameObject } from "./Component";
import { download_file } from "../engine/engine_web_api";
import { getLoader } from "../engine/engine_gltf";
import { InstantiateOptions } from "../engine/engine_gameobject";

const debug = utils.getParam("debugavatar");

export class AvatarModel {
    root: THREE.Object3D;
    head: THREE.Object3D;
    leftHand: THREE.Object3D | null;
    rigthHand: THREE.Object3D | null;


    get isValid(): boolean {
        return this.head !== null && this.head !== undefined;
    }

    constructor(root: THREE.Object3D, head: THREE.Object3D, leftHand: THREE.Object3D | null, rigthHand: THREE.Object3D | null) {
        this.root = root;
        this.head = head;
        this.leftHand = leftHand;
        this.rigthHand = rigthHand;
        this.root?.traverse(h => h.layers.set(2));
        // this.head?.traverse(h => h.layers.set(2));
        // this.leftHand?.traverse(h => h.layers.set(2));
        // this.rigthHand?.traverse(h => h.layers.set(2));
    }
}

export class AvatarLoader {

    private readonly avatarRegistryUrl: string | null = null;
    // private loader: GLTFLoader | null;
    // private avatarModelCache: Map<string, AvatarModel | null> = new Map<string, AvatarModel | null>();

    public async getOrCreateNewAvatarInstance(context: Context, avatarId: string | THREE.Object3D): Promise<AvatarModel | null> {

        if (!avatarId) {
            console.error("Can not create avatar: failed to provide id or root object");
            return null;
        }

        let root: THREE.Object3D | null = null;
        if (typeof avatarId === "string") {
            root = await this.loadAvatar(context, avatarId);
            if (!root) {
                const opts = new InstantiateOptions();
                // opts.parent = context.scene.uuid;
                root = GameObject.instantiate(utils.tryFindObject(avatarId, context.scene), opts);
            }
        }
        else root = avatarId;

        if (!root) {
            return null;
        }
        const model = this.findAvatar(root);
        // model.assignRandomColors();
        // this.cacheModel(avatarId, model);

        if (model.isValid) {
            if (debug)
                console.log("[Custom Avatar] valid config", avatarId, debug ? model : "");
            return model;
        }
        else {
            console.warn("[Custom Avatar] config isn't valid", avatarId, debug ? model : "");
            return null;
        }
    }


    private async loadAvatar(context: Context, avatarId: string): Promise<THREE.Object3D | null> {

        console.assert(avatarId !== undefined && avatarId !== null && typeof avatarId === "string", "Avatar id must not be null");
        if (avatarId.length <= 0) return null;
        if (!avatarId) return null;

        if (debug)
            console.log("[Custom Avatar] " + avatarId + ", loading...");
        // should probably be done by the server?!
        if (!avatarId.endsWith(".glb"))
            avatarId += ".glb";


        // for the demo we use the storage backend we uploaded the avatar to (by file drop)
        if (this.avatarRegistryUrl === null) {
            // try loading avatar from local file
            const download_res = await fetch("./" + avatarId);
            let bin: ArrayBuffer | null = null;
            if (download_res.ok) {
                const blob = await download_res.blob();
                if (blob) bin = await blob.arrayBuffer();
            }
            if (!bin) {
                bin = await download_file(avatarId, avatarId, 0, "no url here go away", true);
                if (!bin) return null;
            }

            const gltf = await getLoader().parseSync(context, bin, null!, 0);
            return gltf?.scene ?? null;
        }


        // if (this.avatarModelCache.has(avatarId)) {
        //     console.log("[Custom Avatar] " + avatarId + ", found in cache");
        //     return new Promise((res, _) => {
        //         var model = this.avatarModelCache.get(avatarId)?.createNewInstance();
        //         res(model);
        //     });
        // }
        // return null;

        const loader = new GLTFLoader();
        loaders.addDracoAndKTX2Loaders(loader, context);

        // TODO: cache binary (fetch finary from model gallery and use binary method)
        return new Promise((resolve, _reject) => {
            const url = this.avatarRegistryUrl + "/" + avatarId;
            loader.load(url,
                async gltf => {
                    await getLoader().createBuiltinComponents(context, url, gltf, null, undefined);
                    resolve(gltf.scene);
                },
                progress => {
                    if (debug)
                        console.log("[Custom Avatar] " + (progress.loaded / progress.total * 100) + '% loaded of ' + (progress.total / 1024) + "kB");
                },
                error => {
                    console.error("[Custom Avatar] " + "Error when loading: " + error);
                    resolve(null);
                }
            );
        });
    }

    private cacheModel(_id: string, _model: AvatarModel) {
        // this.avatarModelCache.set(id, model);
    }

    // TODO this should be burned to the ground once 🤞 we have proper extras that define object relations.
    private findAvatar(obj: THREE.Object3D): AvatarModel {

        const root: THREE.Object3D = obj;
        let searchIn = root;
        // some GLTFs have a "scene" root it seems, others don't, we skip the root here if there's only one child
        if (searchIn.children.length == 1)
            searchIn = obj.children[0];
        let head = this.findAvatarPart(searchIn, ["head"]);

        const leftHand = this.findAvatarPart(searchIn, ["left", "hand"]);
        const rightHand = this.findAvatarPart(searchIn, ["right", "hand"]);

        if (!head) {
            // very last fallback, entire root is used as head
            head = root;

            // normalize size, if the object isn't properly setup the scale might be totally off
            const boundsSize = new THREE.Vector3();
            new THREE.Box3().setFromObject(head).getSize(boundsSize);
            const maxAxis = Math.max(boundsSize.x, boundsSize.y, boundsSize.z);
            console.warn("[Custom Avatar] " + "Normalizing head scale, it's too big: " + maxAxis + " meters! Should be < 0.3m");
            if (maxAxis > 0.3) {
                head.scale.multiplyScalar(1.0 / maxAxis * 0.3);
            }
        }

        const model = new AvatarModel(root, head, leftHand, rightHand);
        return model;
    }


    private findAvatarPart(obj: THREE.Object3D, searchString: string[]): THREE.Object3D | null {

        const name = obj.name.toLowerCase();
        let matchesAll = true;
        for (const str of searchString) {
            if (!matchesAll) break;
            if (name.indexOf(str) === -1)
                matchesAll = false;
        }
        if (matchesAll) return obj;

        if (obj.children) {
            for (const child of obj.children) {
                const found = this.findAvatarPart(child, searchString);
                if (found) return found;
            }
        }
        return null;
    }

    private handleCustomAvatarErrors(response) {
        if (!response.ok) {
            throw Error(response.statusText);
        }
        return response;
    }
}
