// @needle-tools/engine/src/engine-components/ParticleSystemSubEmitter.ts
import { Behavior, Particle, EmissionState, ParticleSystem, ParticleEmitter } from "three.quarks";
import { Vector3, Quaternion, Matrix4 } from "three";
import { IParticleSystem } from "./ParticleSystemModules";
import { CircularBuffer } from "../engine/engine_utils";

const VECTOR_ONE = new Vector3(1, 1, 1);
const VECTOR_Z = new Vector3(0, 0, 1);
const $emitterMatrix = Symbol("emitterMatrix");

export class ParticleSubEmitter implements Behavior {

    type = "NeedleParticleSubEmitter";

    //private matrix_ = new Matrix4();
    private q_ = new Quaternion();
    private v_ = new Vector3();
    private v2_ = new Vector3();


    private _emitterMatrix: Matrix4 = new Matrix4();
    private _circularBuffer: CircularBuffer<Matrix4>;

    constructor(
        private system: IParticleSystem,
        private particleSystem: ParticleSystem,
        private subSystem: IParticleSystem,
        public subParticleSystem?: ParticleSystem
    ) {
        if (this.subParticleSystem && this.subParticleSystem) {
            this.subParticleSystem.onlyUsedByOther = true;
        }
        const maxMatrices = 1000;
        this._circularBuffer = new CircularBuffer(() => new Matrix4(), maxMatrices)
    }

    clone(): Behavior {
        throw new Error("Method not implemented.");
    }

    initialize(particle: Particle): void {
        particle.emissionState = {
            burstIndex: 0,
            burstWaveIndex: 0,
            time: 0,
            waitEmiting: 0,
            // matrix: new Matrix4(),
        } as EmissionState;
        // particle[$emitterMatrix] = new Matrix4();
        this._emitterMatrix.copy(this.subSystem.matrixWorld).invert().premultiply(this.system.matrixWorld)
    }

    update(particle: Particle, delta: number): void {
        if (this.subSystem.currentParticles >= this.subSystem.main.maxParticles)
            return;
        if (!this.subParticleSystem || !particle.emissionState)
            return;

        // TODO: figure out how to re-use matrices
        const m = new Matrix4();// this._circularBuffer.get();// new Matrix4();// particle[$emitterMatrix];

        m.set(
            1, 0, 0, particle.position.x,
            0, 1, 0, particle.position.y,
            0, 0, 1, particle.position.z,
            0, 0, 0, 1
        );

        if (!this.particleSystem.worldSpace) {
            m.multiplyMatrices(this._emitterMatrix, m)
        }

        this.subParticleSystem!.emit(delta, particle.emissionState!, m);
    }

    frameUpdate(_delta: number): void {
    }

    toJSON(): any {
    }
}
