// @needle-tools/engine/lib/engine/extensions/NEEDLE_lightmaps.js
import { FloatType, HalfFloatType, sRGBEncoding } from "three";
import { resolveReferences } from "./extension_utils";
import { getParam } from "../engine_utils";
// the lightmap extension is aimed to also export export skyboxes and custom reflection maps
// should we rename it?
// should we split it into multiple extensions?
export const EXTENSION_NAME = "NEEDLE_lightmaps";
const debug = getParam("debuglightmapsextension");
export var LightmapType;
(function (LightmapType) {
    LightmapType[LightmapType["Lightmap"] = 0] = "Lightmap";
    LightmapType[LightmapType["Skybox"] = 1] = "Skybox";
    LightmapType[LightmapType["Reflection"] = 2] = "Reflection";
})(LightmapType || (LightmapType = {}));
export class NEEDLE_lightmaps {
    get name() {
        return EXTENSION_NAME;
    }
    parser;
    registry;
    source;
    constructor(parser, reg, source) {
        this.parser = parser;
        this.registry = reg;
        this.source = source;
    }
    afterRoot(_result) {
        const extensions = this.parser.json.extensions;
        if (extensions) {
            const ext = extensions[EXTENSION_NAME];
            if (ext) {
                const arr = ext.textures;
                if (!arr?.length) {
                    return null;
                }
                if (debug)
                    console.log(ext);
                return new Promise(async (res, _rej) => {
                    const dependencies = [];
                    for (const entry of arr) {
                        if (entry.pointer) {
                            const res = resolveReferences(this.parser, entry.pointer).then(res => {
                                const tex = res;
                                if (tex?.isTexture) {
                                    if (!this.registry)
                                        console.log(LightmapType[entry.type], entry.pointer, tex);
                                    else {
                                        // TODO this is most likely wrong for floating point textures
                                        if (entry.type !== LightmapType.Lightmap)
                                            tex.encoding = sRGBEncoding;
                                        // not sure why, but seems EXR-loaded float textures need to be flipped
                                        if (entry.type === LightmapType.Skybox) {
                                            if (tex.type == FloatType || tex.type == HalfFloatType)
                                                tex.flipY = true;
                                        }
                                        this.registry.registerTexture(this.source, entry.type, tex, entry.index);
                                    }
                                }
                            });
                            dependencies.push(res);
                        }
                    }
                    await Promise.all(dependencies);
                    res();
                });
            }
        }
        return null;
    }
}
