// @needle-tools/engine/lib/engine-components/EventTrigger.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { serializable } from "../engine/engine_serialization";
import { EventList } from "./EventList";
import { Behaviour } from "./Component";
import { EventType } from "./EventType";
class TriggerEvent {
    eventID;
    callback;
}
__decorate([
    serializable()
], TriggerEvent.prototype, "eventID", void 0);
__decorate([
    serializable(EventList)
], TriggerEvent.prototype, "callback", void 0);
export class EventTrigger extends Behaviour {
    triggers;
    invoke(type) {
        if (!this.triggers)
            return;
        for (const trigger of this.triggers) {
            if (trigger.eventID === type) {
                trigger.callback.invoke();
            }
        }
    }
    onPointerClick(_) {
        this.invoke(EventType.PointerClick);
    }
    onPointerEnter(_) {
        this.invoke(EventType.PointerEnter);
    }
    onPointerExit(_) {
        this.invoke(EventType.PointerExit);
    }
    onPointerDown(_) {
        this.invoke(EventType.PointerDown);
    }
    onPointerUp(_) {
        this.invoke(EventType.PointerUp);
    }
}
__decorate([
    serializable(TriggerEvent)
], EventTrigger.prototype, "triggers", void 0);
