// @needle-tools/engine/lib/engine-components/XRFlag.js
import { Behaviour, GameObject } from "./Component";
import * as utils from "../engine/engine_utils";
const debug = utils.getParam("debugflags");
export var XRStateFlag;
(function (XRStateFlag) {
    XRStateFlag[XRStateFlag["Never"] = 0] = "Never";
    XRStateFlag[XRStateFlag["Browser"] = 1] = "Browser";
    XRStateFlag[XRStateFlag["AR"] = 2] = "AR";
    XRStateFlag[XRStateFlag["VR"] = 4] = "VR";
    XRStateFlag[XRStateFlag["FirstPerson"] = 8] = "FirstPerson";
    XRStateFlag[XRStateFlag["ThirdPerson"] = 16] = "ThirdPerson";
    XRStateFlag[XRStateFlag["All"] = 4294967295] = "All";
})(XRStateFlag || (XRStateFlag = {}));
// console.log(XRStateFlag);
export class XRState {
    static Global = new XRState();
    Mask = XRStateFlag.Browser | XRStateFlag.ThirdPerson;
    Has(state) {
        const res = (this.Mask & state);
        return res !== 0;
    }
    Set(state) {
        if (debug)
            console.warn("Set XR flag state to", state);
        this.Mask = state;
        XRFlag.Apply();
    }
    Enable(state) {
        this.Mask |= state;
        XRFlag.Apply();
    }
    Disable(state) {
        this.Mask &= ~state;
        XRFlag.Apply();
    }
    Toggle(state) {
        this.Mask ^= state;
        XRFlag.Apply();
    }
    EnableAll() {
        this.Mask = 0xffffffff | 0;
        XRFlag.Apply();
    }
    DisableAll() {
        this.Mask = 0;
        XRFlag.Apply();
    }
}
export class XRFlag extends Behaviour {
    static registry = [];
    static Apply() {
        for (const r of this.registry)
            r.UpdateVisible(XRState.Global);
    }
    static firstApply;
    static buffer = new XRState();
    awake() {
        XRFlag.registry.push(this);
    }
    onEnable() {
        if (!XRFlag.firstApply) {
            XRFlag.firstApply = true;
            XRFlag.Apply();
        }
    }
    onDestroy() {
        const i = XRFlag.registry.indexOf(this);
        if (i >= 0)
            XRFlag.registry.splice(i, 1);
    }
    visibleIn;
    get isOn() { return this.gameObject.visible; }
    UpdateVisible(state = null) {
        // XR flags set visibility of whole hierarchy which is like setting the whole object inactive
        // so we need to ignore the enabled state of the XRFlag component
        // if(!this.enabled) return;
        let res = undefined;
        const flag = state;
        if (flag && typeof flag === "number") {
            console.assert(typeof flag === "number", "XRFlag.UpdateVisible: state must be a number", flag);
            if (debug)
                console.log(flag);
            XRFlag.buffer.Mask = flag;
            state = XRFlag.buffer;
        }
        const st = state;
        if (st) {
            if (debug)
                console.warn(this.name, "use passed in mask", st.Mask, this.visibleIn);
            res = st.Has(this.visibleIn);
        }
        else {
            if (debug)
                console.log(this.name, "use global mask");
            XRState.Global.Has(this.visibleIn);
        }
        if (res === undefined)
            return;
        if (res) {
            if (debug)
                console.log(this.name, "is visible", this.gameObject.uuid);
            // this.gameObject.visible = true;
            GameObject.setActive(this.gameObject, true);
        }
        else {
            if (debug)
                console.log(this.name, "is not visible", this.gameObject.uuid);
            const isVisible = this.gameObject.visible;
            if (!isVisible)
                return;
            this.gameObject.visible = false;
            // console.trace("DISABLE", this.name);
            // GameObject.setActive(this.gameObject, false);
        }
    }
}
